File: src/NaiveVoting.java
import java.util.ArrayList;

/**
 * Created by RobbinNi on 26/11/2017.
 */

public class NaiveVoting {

    public static final int LEADER = 0;

    private final Environment e;

    private int input, output;

    private int[] cnt = new int[2];

    public NaiveVoting(Environment e) {
        this.e = e;
    }

    public void runNode() {
        int round = e.getCurRound();
        if (round == 0) {
            if (e.getID(this) == LEADER) {
                input = e.getInput(this);
                e.putBroadcast(this, new Message(input));
            } else {
                input = -1;
            }
            cnt[0] = cnt[1] = 0;
        } else if (round == 1) {
            ArrayList<Message> msgs = e.getInputMessage(this);
            int b = 0;
            for (Message m : msgs) {
                b = m.getContent();
            }
            e.putBroadcast(this, new Message(b));
        } else if (round == 2) {
            ArrayList<Message> msgs = e.getInputMessage(this);
            for (Message m : msgs) {
                int b = m.getContent();
                if (b == 0 || b == 1) {
                    cnt[b]++;
                }
            }
        } else if (round == 3) {
            System.out.println(e.getID(this) + " : " + cnt[0] + " " + cnt[1]);
            int bar = e.getN() * 2 / 3;
            if (cnt[0] >= bar) {
                output = 0;
            } else if (cnt[1] >= bar) {
                output = 1;
            } else {
                output = -1;
            }
            e.putOutput(this, output);
        } else {
            throw new RuntimeException();
        }
    }
}

File: src/Protocols/NaiveVoting.py
from Messages.Message import Message


class NaiveVoting:
    name = "Naive Voting Protocol"
    SENDER = 0

    def __init__(self, env, pki):
        self.env = env
        self.pki = pki
        self.pki.register(self)
        self.input = None

    def run_node(self):
        round = self.env.get_round()
        myid = self.env.get_id(self)
        if round == 0:
            if self.env.get_id(self) == NaiveVoting.SENDER:
                self.input = self.env.get_input(self)
                self.env.put_broadcast(self, self.pki.sign(
                    self, Message(myid, self.input)))
        elif round == 1:
            msgs = self.env.get_input_msgs(self)
            for msg in msgs:
                if self.pki.verify(msg) and msg.get_sender() == NaiveVoting.SENDER and (
                        msg.get_extraction() == 0 or msg.get_extraction() == 1):
                    self.env.put_broadcast(
                        self, self.pki.sign(self, Message(myid, msg)))
        elif round == 2:
            msgs = self.env.get_input_msgs(self)
            seen = set()
            cnt = [0, 0]
            for msg in msgs:
                if self.pki.verify(msg) and (msg.get_sender() not in seen) and (
                        msg.get_extraction() == 0 or msg.get_extraction() == 1):
                    seen.add(msg.get_sender())
                    cnt[msg.get_extraction()] += 1
            bar = self.env.get_n() * 2 / 3
            if cnt[0] >= bar:
                self.env.put_output(self, 0)
            elif cnt[1] >= bar:
                self.env.put_output(self, 1)
            else:
                self.env.put_output(self, None)
        else:
            raise RuntimeError
